libraries/AP_PiccoloCAN/AP_PiccoloCAN.cpp

```diff
@@ -0,0 +1,624 @@
+{
+    _driver_index = driver_index;
+
+    debug_can(2, "PiccoloCAN: starting init\n\r");
+
+    if (_initialized) {
+        debug_can(1, "PiccoloCAN: already initialized\n\r");
+        return;
+    }
+
+    AP_HAL::CANManager* can_mgr = hal.can_mgr[driver_index];
+
+    if (can_mgr == nullptr) {
+        debug_can(1, "PiccoloCAN: no mgr for this driver\n\r");
+        return;
+    }
+
+    if (!can_mgr->is_initialized()) {
+        debug_can(1, "PiccoloCAN: mgr not initialized\n\r");
+        return;
+    }
+
+    _can_driver = can_mgr->get_driver();
+
+    if (_can_driver == nullptr) {
+        debug_can(1, "PiccoloCAN: no CAN driver\n\r");
+        return;
+    }
+
+    // start calls to loop in separate thread
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_PiccoloCAN::loop, void), _thread_name, 4096, AP_HAL::Scheduler::PRIORITY_MAIN, 1)) {
+        debug_can(1, "PiccoloCAN: couldn't create thread\n\r");
+        return;
+    }
+
+    _initialized = true;
+
+    snprintf(_thread_name, sizeof(_thread_name), "PiccoloCAN_%u", driver_index);
+
+    debug_can(2, "PiccoloCAN: init done\n\r");
+}
+
+// loop to send output to CAN devices in background thread
+void AP_PiccoloCAN::loop()
+{
+    uavcan::CanFrame txFrame;
+    uavcan::CanFrame rxFrame;
+
+    // How often to transmit CAN messages (milliseconds)
+#define CMD_TX_PERIOD 10
+
+    uint16_t txCounter = 0;
+
+    // CAN Frame ID components
+    uint8_t frame_id_group;     // Piccolo message group
+    uint16_t frame_id_device;   // Device identifier
+
+    uavcan::MonotonicTime timeout;
+
+    while (true) {
+
+        if (!_initialized) {
+            debug_can(2, "PiccoloCAN: not initialized\n\r");
+            hal.scheduler->delay_microseconds(10000);
+            continue;
+        }
+
+        timeout = uavcan::MonotonicTime::fromUSec(AP_HAL::micros64() + 250);
+
+        // 1ms loop delay
+        hal.scheduler->delay_microseconds(1 * 1000);
+
+        // Transmit CAN commands at regular intervals
+        if (txCounter++ > CMD_TX_PERIOD) {
+
+            txCounter = 0;
+
+            // Transmit ESC commands
+            send_esc_messages();
+        }
+
+        // Look for any message responses on the CAN bus
+        while (read_frame(rxFrame, timeout)) {
+            frame_id_group = (rxFrame.id >> 24) & 0x1F;
+            frame_id_device = (rxFrame.id >> 8) & 0xFF;
+
+            // Only accept extended messages
+            if ((rxFrame.id & uavcan::CanFrame::FlagEFF) == 0) {
+                continue;
+            }
+
+            switch (MessageGroup(frame_id_group)) {
+            // ESC messages exist in the ACTUATOR group
+            case MessageGroup::ACTUATOR:
+
+                switch (ActuatorType(frame_id_device)) {
+                case ActuatorType::ESC:
+                    if (handle_esc_message(rxFrame)) {
+                        // Returns true if the message was successfully decoded
```

Comment:
is it worth showing an err msg if we can't decode? How likely is that with this CAN protocol?

Reply:
There are plenty of other messages which are contained within the ESC protocol space but not useful for integration here. So the function *handle_esc_message* may return false if there is a message which is "valid" but it doesn't have a matching decode function here.